keps/0017-pipe-tasks.md

```diff
@@ -0,0 +1,122 @@
+---
+kep-number: 17
+title: Pipe Tasks
+authors:
+  - "@zen-dog"
+owners:
+  - "@zen-dog"
+editor: TBD
+creation-date: 2019-09-12
+last-updated: 2019-09-12
+status: provisional
+---
+
+# Pipe Tasks
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Limitations](#limitations)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints-optional)
+* [Alternatives](#alternatives)
+
+## Summary
+
+Developing complicated operators often require generating files in one task and reusing them in a later one. A common example is generating custom certificates/dynamic configuration files in the bootstrap task and using them in the later deployment step of the service. This kep describes how KUDO could help operator developers by making this task seamless and automated.
+
+## Motivation
+
+We want to ease the development of operators with complicated life-cycles by enabling to pipe files from one task to another. While this can be achieved in a manual way, involving custom scripts, this is exactly the kind of "magic glue" that KUDO could provide to the operator developers to ease the development process.
+
+### Goals
+
+The goal of this kep is to describe application examples, limitations and ways to implement passing files between tasks. We aim to help operator developers pipe small files (<1Mb) like certificates and dynamically generated configuration which can be referenced later as ConfigMaps and Secrets.
+
+### Non-Goals
+
+Allowing to pipe all kind of files between tasks requires a general-purpose storage solution (e.g. S3) available in the cluster. While certainly possible, this is not the goal of this kep.
+
+## Proposal
+
+This section describes how pipe tasks and files they produce are configured in the operator. Here is a task definition that produces a file that will be stored as a Secret, referenced by `{{.Pipes.Certificate}}` key:
+```yaml
+steps:
+    - name: bootstrap
+    tasks:
+        - gencert
+        - pipe:
+            - file: /usr/share/MyKey.key
+              kind: Secret # ConfigMap
+              key: {{.Pipes.Certificate}}
```

Comment:
all the pipes will use the `.Pipes.Certificate`? I thought that is a variable so it cannot be backed by multiple values... or that is unique per pipe and that is the variable to which kudo will generate the one unique value for this concrete pipe?

Reply:
instead of autogenerated name we can always generathe the name from the plan/phase/step combination, I think TeamCity or other do that too. So you can reference your artifact by somethng like `deploy.bootstrap.gencert.pipe` ... of course teamcity makes this easier because it helps you discover what is availably by hinting you while typing in the UI but we could ultimately make this also easier by some CLI tools.

Reply:
{{.Pipes.Certificate}} is a variable, set by a pipe task **once** (we might add validation to prevent mistakes), where the actual string (e.g. secret-name) will be generated by KUDO behind the scenes. I like the idea of hierarchical names e.g. `deploy.bootstrap.gencert.pipe.certificate-#hash` - makes it immediately clear, where it belongs too. But I thought about the name string composition as an implementation detail where the user would use the variable.

Reply:
I've added these points to the KEP.